src/v/storage/segment_index.h

```diff
@@ -29,6 +29,71 @@
 
 namespace storage {
 
+// clang-format off
+// this truth table shows which timestamps gets used as retention_timestamp
+//
+// use_broker_ts  has_broker_ts  ignore_future_ts  has_alternative_to_future_ts  retention_ts
+//                               (likely false)
+// TRUE           TRUE           TRUE              TRUE                          broker_timestamp
+// TRUE           TRUE           TRUE              FALSE                         broker_timestamp
+// TRUE           TRUE           FALSE             TRUE                          broker_timestamp
+// TRUE           TRUE           FALSE             FALSE                         broker_timestamp             // new segment, new cluster
+// TRUE           FALSE          TRUE              TRUE                          segment_index::_retention_ms // buggy old segment, new cluster
+// TRUE           FALSE          TRUE              FALSE                         max_timestamp
+// TRUE           FALSE          FALSE             TRUE                          max_timestamp
+// TRUE           FALSE          FALSE             FALSE                         max_timestamp                // old segment, new cluster
+// FALSE          TRUE           TRUE              TRUE                          segment_index::_retention_ms
+// FALSE          TRUE           TRUE              FALSE                         max_timestamp
+// FALSE          TRUE           FALSE             TRUE                          max_timestamp
+// FALSE          TRUE           FALSE             FALSE                         max_timestamp                // new segment, upgraded cluster
+// FALSE          FALSE          TRUE              TRUE                          segment_index::_retention_ms // buggy old segments
+// FALSE          FALSE          TRUE              FALSE                         max_timestamp
+// FALSE          FALSE          FALSE             TRUE                          max_timestamp
+// FALSE          FALSE          FALSE             FALSE                         max_timestamp                // old segment, upgraded cluster
+// clang-format on
+
+// this struct is meant to be a local copy of the feature
+// broker_time_based_retention and configuration property
+// storage_ignore_timestamps_in_future_secs
+struct time_based_retention_cfg {
+    bool use_broker_time;
+    bool use_escape_hatch_for_timestamps_in_the_future;
+
+    static auto make(features::feature_table const& ft)
+      -> time_based_retention_cfg {
+        return {
+          .use_broker_time = ft.is_active(
+            features::feature::broker_time_based_retention),
+          .use_escape_hatch_for_timestamps_in_the_future
+          = config::shard_local_cfg()
+              .storage_ignore_timestamps_in_future_sec()
+              .has_value(),
+        };
+    }
+
+    /// this function is the codification of the table above
+    constexpr auto compute_retention_ms(
```

Comment:
a force of habit when I write a method in the header. In the first version, it was truly constexpr but then I wired in feature_table and configuration so now it's extremely unlikely that it will ever be constexpr. I'll change it.